File: code/data_structures/src/tree/space_partitioning_tree/quad_tree/Python Code with Visualization/QuadTree.py
from tkinter import *


class Rectangle:
    def __init__(self, x, y, h, w):
        self.x = x
        self.y = y
        self.h = h
        self.w = w


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y


class QuadTree:
    def __init__(self, boundary, bucket):
        self.boundary = boundary
        self.bucket = bucket
        self.arr = []
        self.NorthEast = None
        self.NorthWest = None
        self.SouthEast = None
        self.SouthWest = None
        self.divided = False

    def contains(self, point):
        if (((point.x >= self.boundary.x - self.boundary.w) and (point.x <= self.boundary.x + self.boundary.w)) and
                ((point.y <= self.boundary.y + self.boundary.h) and (point.y >= self.boundary.y - self.boundary.h))):
            return True
        return False

    def subdivide(self):
        nE = Rectangle(self.boundary.x + (self.boundary.w / 2), self.boundary.y + (self.boundary.h / 2),
                       self.boundary.w / 2, self.boundary.h / 2)
        nW = Rectangle(self.boundary.x - (self.boundary.w / 2), self.boundary.y + (self.boundary.h / 2),
                       self.boundary.w / 2, self.boundary.h / 2)
        sE = Rectangle(self.boundary.x + (self.boundary.w / 2), self.boundary.y - (self.boundary.h / 2),
                       self.boundary.w / 2, self.boundary.h / 2)
        sW = Rectangle(self.boundary.x - (self.boundary.w / 2), self.boundary.y - (self.boundary.h / 2),
                       self.boundary.w / 2, self.boundary.h / 2)
        self.NorthEast = QuadTree(nE, self.bucket)
        self.NorthWest = QuadTree(nW, self.bucket)
        self.SouthEast = QuadTree(sE, self.bucket)
        self.SouthWest = QuadTree(sW, self.bucket)

    def insert(self, point):
        if not self.contains(point):
            return False

        if len(self.arr) < self.bucket:
            self.arr.append(point)
            return True

        if not self.divided:
            self.subdivide()
            self.divided = True

        if self.NorthEast.insert(point) is not None:
            return True
        if self.NorthWest.insert(point) is not None:
            return True
        if self.SouthEast.insert(point) is not None:
            return True
        if self.SouthWest.insert(point) is not None:
            return True

        return False

    def visualize(self, pointsList):
        root = Tk()
        canvas = Canvas(root, height=700, width=700, bg='#fff')
        canvas.pack()
        for point in pointsList:
            canvas.create_oval(point.x - 3, point.y - 3, point.x + 3, point.y + 3)
        self.paint(canvas)
        root.mainloop()

    def paint(self, canvas):
        canvas.create_rectangle(self.boundary.x - self.boundary.w, self.boundary.y + self.boundary.h,
                                self.boundary.x + self.boundary.w, self.boundary.y - self.boundary.h, outline='#fb0')

        if self.divided:
            self.NorthEast.paint(canvas)
            self.NorthWest.paint(canvas)
            self.SouthEast.paint(canvas)
            self.SouthWest.paint(canvas)

File: code/data_structures/src/tree/space_partitioning_tree/quad_tree/quad_tree.cpp
// Part of Cosmos by OpenGenus Foundation
#include "iostream"
#include "vector"

#define ll long long
using namespace std;

struct Rectangle {
    float x;
    float y;
    float h;
    float w;
};

struct Point {
    float x;
    float y;
};


class QuadTree {
    Rectangle boundary{};
    vector<Point> arr;
    QuadTree *NorthEast;
    QuadTree *NorthWest;
    QuadTree *SouthEast;
    QuadTree *SouthWest;
    bool divided;
    int bucket;
public:
    QuadTree(Rectangle b, int lim) {
        boundary = b;
        bucket = lim; // Number of Points that can be included in one quadrant.
        divided = false;
        NorthEast = nullptr;
        NorthWest = nullptr;
        SouthEast = nullptr;
        SouthWest = nullptr;
    }

    [[nodiscard]] bool contains(Point point) const {
        if (((point.x >= boundary.x - boundary.w) and (point.x <= boundary.x + boundary.w)) and
            ((point.y <= boundary.y + boundary.h) and (point.y >= boundary.y - boundary.h))) {
            return true;
        }
        return false;
    }


    void SubDivide() {
        float x = boundary.x;
        float y = boundary.y;
        float h = boundary.h;
        float w = boundary.w;
        Rectangle nE{x + (w / 2), y + (h / 2), w / 2, h / 2};
        NorthEast = new QuadTree(nE, bucket);
        Rectangle nW{x - (w / 2), y + (h / 2), w / 2, h / 2};
        NorthWest = new QuadTree(nW, bucket);
        Rectangle sE{x + (w / 2), y - (h / 2), w / 2, h / 2};
        SouthEast = new QuadTree(sE, bucket);
        Rectangle sW{x - (w / 2), y - (h / 2), w / 2, h / 2};
        SouthWest = new QuadTree(sW, bucket);
    }

    bool insert(Point point) {
        if (!contains(point)) {
            return false;
        }

        if (arr.size() < bucket) {
            arr.push_back(point);
            return true;
        }

        if (!divided) {
            SubDivide();
            divided = true;
        }
        if (NorthEast->insert(point)) return true;
        if (NorthWest->insert(point)) return true;
        if (SouthEast->insert(point)) return true;
        if (SouthWest->insert(point)) return true;

        return false;
    }

    void traverse(QuadTree qt) {
        if (qt.arr.empty()) {
            return;
        }
        cout << qt.arr[0].x << " " << qt.arr[0].y << "\n";
        if (qt.NorthEast != nullptr) {
            QuadTree temp3 = *qt.NorthEast;
            traverse(temp3);
        }
        if (qt.NorthWest != nullptr) {
            QuadTree temp4 = *qt.NorthWest;
            traverse(temp4);
        }
        if (qt.SouthWest != nullptr) {
            QuadTree temp1 = *qt.SouthWest;
            traverse(temp1);
        }
        if (qt.SouthEast != nullptr) {
            QuadTree temp2 = *qt.SouthEast;
            traverse(temp2);
        }
    }


};

int main() {
    Rectangle boundary{0, 0, 500,
                       500}; //Any inserted coordinate => (x, y) should satisfy => (boundary.x - boundary.w <= x <= boundary.x + boundary.w) and (boundary.y - boundary.h <= y <= boundary.y + boundary.h)
    QuadTree QT = QuadTree(boundary, 1);
    QT.insert(Point{0, 0});
    QT.insert(Point{1, 1});
    QT.insert(Point{-1, -1});
    QT.insert(Point{1, -1});
    QT.insert(Point{-1, 1});
    QT.insert(Point{2, 2});
    QT.insert(Point{-2, 2});
    QT.insert(Point{2, -2});
    QT.insert(Point{-2, -2});
    QT.traverse(QT);
    return 0;
}
